node-graph/gstd/src/image_color_palette.rs

```diff
@@ -0,0 +1,44 @@
+use graphene_core::raster::ImageFrame;
+use graphene_core::Color;
+use graphene_core::Node;
+
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct ImageColorPaletteNode;
+
+#[node_macro::node_fn(ImageColorPaletteNode)]
+fn image_color_palette(frame: ImageFrame<Color>) -> Vec<Color> {
+	const GRID_SIZE: u8 = 3;
+	const NUM_BUCKETS: usize = (GRID_SIZE * GRID_SIZE * GRID_SIZE) as usize;
+
+	let mut histogram: Vec<u32> = vec![0; NUM_BUCKETS];
+
+	for color in frame.image.data.iter() {
+		let r = (color.r() * (GRID_SIZE - 1) as f32) as u8;
+		let g = (color.g() * (GRID_SIZE - 1) as f32) as u8;
+		let b = (color.b() * (GRID_SIZE - 1) as f32) as u8;
+
+		let bucket = (r * GRID_SIZE * GRID_SIZE + g * GRID_SIZE + b) as usize;
+
+		histogram[bucket] += 1;
+	}
+
+	let mut sorted = histogram.iter().enumerate().collect::<Vec<(usize, &u32)>>();
+
+	sorted.sort_by(|(_, count1), (_, count2)| count2.cmp(count1));
+
+	let top_buckets = &sorted[..10];
+
+	let palette = top_buckets
+		.iter()
+		.map(|(i, _)| {
+			let r = (i / (GRID_SIZE as usize * GRID_SIZE as usize)) as f32 / (GRID_SIZE - 1) as f32;
+			let g = ((i / GRID_SIZE as usize) % GRID_SIZE as usize) as f32 / (GRID_SIZE - 1) as f32;
+			let b = (i % GRID_SIZE as usize) as f32 / (GRID_SIZE - 1) as f32;
+
+			return Color::from_rgbaf32(r as f32 / GRID_SIZE as f32, g as f32 / GRID_SIZE as f32, b as f32 / GRID_SIZE as f32, 1.0).unwrap();
+		})
+		.collect::<Vec<Color>>();
+
+	return palette;
+}
```

Comment:
What could I do to create some kind of unit test for this function to guarantee that its implementation is right? I have tried a “traditional” manner, but many random errors appeared.

Reply:
~it is still not correctly~